Fix part of #1861: [RunAllTests] Update Bazel CI workflow to run all Oppia tests

BenHenning opened the pull request:
Fix part of #1861.

At a high-level, this PR introduces running all of Oppia's Robolectric tests in GitHub Actions via Bazel now that Bazel has shown sufficient maturity while being used to continually build the app binary. In particular:
1. The tests take much longer to run since Bazel's builds are more complicated than on Gradle. This has been partially mitigated by improving parallelization (e.g. running each test as its own action), among other techniques described below. However, even with these optimizations Gradle will still outperform Bazel in the long-term when it comes to running tests. This is a cost we need to accept as a team.
2. One optimization done here is leveraging Bazel's remote caching functionality. However, to ensure the cache remains properly locked down, only branches created directly in this repo can leverage caching functionality. Consequently, forks from contributors will take longer to run through the actions. Despite caching, the overall performance benefits aren't as great as expected: only about 20% of the actions are being restored during build time. This is due to poor reproducibility in the data-binding part of the Android resources build pipeline and will require further investigation & fixes to address. The remote caching functionality is locked down using git-secret. Currently only I have access to the secrets, plus the fake robot account used by the workflow runner.
3. The new actions actually introduce computations for affected targets. This means only the tests relevant to the underlying changes will be run, except if on ``develop`` (e.g. for continuous builds) in which case all tests will run, or if the PR title includes ``[RunAllTests]``. While this is a nice addition, the real benefits won't be realized for a while: the lack of modularization means small changes (e.g. to just 1 app module file) will result in all tests tied to that module to need to re-run, and any changes to Bazel/BUILD files will trigger a re-run of the entire test suite since the Bazel files transitively expand to the whole set of BUILD files due to a circular dependency between domain & app. Changes to WORKSPACE will always trigger a full test re-run for safety.
4. This PR introduces automatic workflow cancellation. This is even more important now that we're running so many actions at once.
5. GitHub will try to run as many actions at once as it can. I've seen times when it only runs 5 simultaneously, and other times where it runs nearly 50. This depends on current active resources across the organization, so we'll need to keep an eye on runtime to see how good/bad it is for contributors.
6. This PR also fixes a few build errors that have crept up in the last few weeks/months.
9. Per (3) this PR is running all of the tests to demonstrate that they work as expected (though given that a Bazel file was changed in this PR, it would be likely that all of the tests would have been run, anyway).
10. This PR includes a few test fixes that are specific to JDK 9 environments for unknown reasons. See https://github.com/oppia/oppia-android/pull/1904#issuecomment-732398008 and comments above it for more context.
11. A new workflow is being introduced to check that all tests passed. This is needed since the required checks in GitHub's settings can't require all test runs to be automatically required since it uses a matrix. Since this can change over time, we can't rely on specifying each test as required. Instead, we will use a static job that runs after all the others as the required check.

For some side context, #1876 required a lot of peripheral changes due to previous changes breaking Bazel test targets. This will help prevent that from happening. This PR will also help improve test reliability since it caught at least 1 test suite that had some timing issues that needed to be fixed that wasn't hit with Gradle (StateFragmentLocalTest).

FYI that #1861 is tracking long-term optimizations for improving Bazel builds in CI beyond those included in this PR.

Follow-up items to complete after this PR is merged:
- [x] 1-5. #2691
- [x] 6. ~Introduce a wiki article explaining how to setup Bazel locally to investigate CI failures (this is going to be a bit rough initially since Bazel 4.0 isn't released yet, so the custom build of Bazel will need to be used in the interim).~ Now available [here](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions).

Note that GitHub seems to also have a nice visualization of the job dependency graph. For specifics, see: https://github.com/oppia/oppia-android/actions/runs/424948426.

BenHenning committed:
Update Bazel CI workflow

Build all Oppia targets in Bazel rather than just the binary target.

BenHenning committed:
Update main.yml

Reset the workflow name so that it doesn't need to be renamed in GitHub settings.

BenHenning commented:
Note that the build will be failing until after #1876 is merged. Switching to draft for now.

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning commented:
#1876 is now merged and this PR is up-to-date so it should pass. Sending it to review.

BenHenning commented:
Hmm, maybe this isn't feasible. CI killed this after an hour run--I wasn't expecting it to take that long.

miaboloix commented:
> Hmm, maybe this isn't feasible. CI killed this after an hour run--I wasn't expecting it to take that long.

Maybe we should try to create various tasks instead of one large task that builds all Bazel targets? I was thinking that since we already upload the Bazel binary we could download it for use in other tasks - maybe creating tasks that build all the Bazel targets for utility, then a task for model, etc. What do you think?

BenHenning commented:
We seem to be able to build the Android binary in ~15 minutes which covers all of the libraries. I think there might be some base overhead to tests that are being shared across the 100 test suites, and CI isn't great at parallel builds. I think maybe building the binary & then sharing those build results with the test run might help: we can actually share all of the libraries this way, and hopefully the tests go by a bit faster. I think this means that even running all the tests with Bazel will exceeds 50 minutes, though, which is definitely problematic.

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning commented:
So I've had some luck locally with setting up remote caching--this seems to decrease build times 2-5x which might make an optional build-all actions check viable. We'll probably want to split up the build based on modules, though, to improve parallelization.

BenHenning committed:
Introduce remote caching in Bazel.

This uses a remote storage service with a local file encrypted using
git-secret to act as the authentication key for Bazel to read & write
artifacts to the service for caching purposes.

BenHenning committed:
Add debug line.

BenHenning committed:
Disable workflows + fix debug line.

BenHenning committed:
More debugging.

BenHenning committed:
More debugging.

BenHenning committed:
Work around GitHub hiding secret since we're debugging.

BenHenning committed:
Use base64 to properly encode newlines in GPG keys.

BenHenning committed:
Remove debug lines before changing back to correct GPG key.

BenHenning committed:
Switch to production key.

BenHenning committed:
Fix env variable reference. Lock-down actions workflows via codeowners.

BenHenning committed:
Install git-secret to default location.

BenHenning committed:
Add details. Debug $PATH.

BenHenning committed:
Fix pathing for git-secret.

BenHenning committed:
Dummy commit to re-trigger actions.

BenHenning committed:
Undo commit to see if this one shows up.

BenHenning committed:
Fix git-secret pathing try 2.

BenHenning committed:
New commit to re-trigger action.

BenHenning committed:
Path debugging.

BenHenning committed:
Workaround to get GitHub to show changes.

BenHenning committed:
Update runner to use Bash.

Reference:
https://github.community/t/github-path-does-not-add-to-the-path/143992.

BenHenning committed:
Restore binary-only build, other builds, and introduce new workflow for
building all Bazel targets.

BenHenning commented:
Good news: the latest run (with caching) took about an hour to build everything. That's much better than the earlier 3 hour attempt that timed out. :) Trying another run (this time with the build-all as a separate workflow) to see if it runs even faster.

The new workflow will be optional since it can't run on forks due to needing access to GitHub secrets. It will be useful to check for runs on develop, though, as a post-merge check.

BenHenning committed:
Remove debug lines.

BenHenning commented:
Less good news: a forced re-run led to much less caching and a longer build time. I'm not totally sure why.

BenHenning committed:
Rename & remove keep_going.

BenHenning committed:
Compute matrix containing all test targets.

This will allow us to distribute parallelization responsibilities partly
to GitHub actions to hopefully have slightly better throughput. See
https://github.blog/changelog/2020-04-15-github-actions-new-workflow-features/
for reference on how this mechanism works.

BenHenning committed:
Simplify, fix some issues, and debug instead of run.

BenHenning commented:
Looking at the debug run, GitHub Actions is actually running far more concurrently than I expected: 
![image](https://user-images.githubusercontent.com/12983742/99749551-9e571b80-2a93-11eb-9814-0cdc0dc9ccf3.png).

While this isn't actually running the expensive tests, it is cloning the ~500mb Bazel repository over and over again. This is actually quite fast, so I'm excited to see how it operates with a real test load.

Note that the plan here is to build the entire Oppia binary with caching, then kick-off the tests in parallel (hopefully to improve build performance, but there are some caching issues as noted in https://github.com/oppia/oppia-android/issues/1861#issuecomment-730793256.

BenHenning committed:
Turn on actual testing.

BenHenning commented:
Okay, marking comment to mark when the CI starts for real testing (other, unrelated tests/builds are currently disabled to emphasize the part we care about checking here).

BenHenning commented:
Half the tests ran in about 37 minutes, but half were cancelled due to exceeding GitHub's actions limit. :) Need to figure out a way to tweak this better.

BenHenning committed:
Lower parallelization since GitHub started cancelling tasks.

BenHenning commented:
Trying again  with max parallelization of 25. This should take longer, but be more resilient to GitHub killing tasks.

BenHenning committed:
Try 15 parallel jobs instead.

BenHenning committed:
Turn off fail-fast instead of limiting parallelization.

Fail fast seems to be the reason why the tests aren't completing, not
quota (since if too many jobs are started, the extras should just be
queued until resources open up).

BenHenning commented:
Looks like the whole workflow ran in about 54 minutes: https://github.com/oppia/oppia-android/actions/runs/373726047. That's by far the fastest we've seen the builds run yet, and this actually including running the test themselves not just building them.

I'm still trying to work out how to conditionally require one job vs. another to pass (since we want to run caching for non-forks and caching for forks).

BenHenning commented:
So I tried locally copying the local Bazel cache & reuse it across subsequent builds (using ``--disk_cache``), but it doesn't seem to offer any benefit. I suspect it's for the same reason remote caching isn't working well--part of the data-binding pipeline isn't reproducible and is wiping out most of the build cache for each subsequent build. We can probably live with this for now, but long-term we'll want to figure out how to fix that so that builds are *much* faster.

BenHenning committed:
Simplify workflow & allow it to be required.

Also, introduce bazelrc file to simplify the CI CLIs interacting with
Bazel.

BenHenning committed:
Add test change to investigate computing affected targets.

BenHenning committed:
Another test change to compute affected targets.

BenHenning committed:
Update workflow to use future script compute_affected_tests.sh.

Also, ignore Bazel directories in Git (to ease local development).

BenHenning committed:
Add script to compute affected targets.

This script is primarily meant to be used for CI.

BenHenning commented:
I'm starting to think that having a dependency on the full Bazel build may not actually be the most efficient. Going to try changing up the workflow a bit.

BenHenning committed:
Execute tests in parallel to build.

This creates a new job to compute affected targets alongside the build.
This may result in the initial build being a bit slower, but subsequent
commits should be fast if remote caching is enabled. This will also
result in better performance for forks that can't use remote caching.

BenHenning committed:
Script clean-ups.

Also, re-trigger actions.

BenHenning committed:
Try to ensure develop branch is available for change analysis.

BenHenning commented:
![image](https://user-images.githubusercontent.com/12983742/99854811-6656e380-2b3a-11eb-870d-8da6f126a343.png)

Running only affected tests is working. :) There are still a few issues:
1) The codebase isn't modularized, so small changes leads to large numbers of tests being re-run
2) Affected targets doesn't properly factor in changes to BUILD files, so I'm working to figure out a way to address that

BenHenning committed:
Add automatic workflow cancellation.

Also, add support to explicitly run all tests if '[RunAllTests]' is in
the PR title.

BenHenning commented:
Incorporated an automatic canceller so that previous runs don't continue using resources when their results aren't needed: https://github.com/oppia/oppia-android/actions/runs/375234456.

BenHenning committed:
Attempt to make conditional matrix computation work.

BenHenning commented:
Restart workflows.

BenHenning committed:
Remove join since it was used incorrectly.

BenHenning commented:
Looks like the new ``[RunAllTests]`` title description is working correctly: it will forcibly run all tests (which is useful in a pinch for cases when the computation for affected tests computes the wrong list).

BenHenning committed:
Add support for testing when Bazel, BUILD, and WORKSPACE files change.

One consequence is the current Bazel file structure is very tight, so
any changes will likely run the whole suite. This will get better over
time.

Also, show test logs for all test runs.

BenHenning committed:
Fix broken tests by adding missing dep library.

BenHenning committed:
Finalize PR.

1. Expand codeowners to include all workflow files.
2. Remove test comments in Kotlin files.
3. Re-enable all workflows.
4. Attempt to fix tests broken on actions but not locally by adding more
thread synchronization points.

BenHenning committed:
Lint fix.

BenHenning commented:
Okay, I think some additional work is needed here: I can't actually enable the run workflow as a status check because GitHub treats each test distinctly due to the matrix strategy. I'll need to introduce some sort of job that depends on all of the previous to run first before passing iff all of the tests passed.

I'll need to follow up on this next week, but after that's addressed this PR should  be ready for review.

BenHenning commented:
I tried reproing the StateFragmentLocalTest failures locally by using a cpu limiter--that didn't seem to help (plus, GitHub actions is running the test pretty close to the same timeframe as my local workstation--only about 30-40% slower).

BenHenning commented:
Aha, I was able to repro 1 of the failures locally by using JDK 9 instead of 8. That's actually really interesting.

BenHenning commented:
Okay, this is quite interesting. After further investigating, it seems like the ViewGroup children are being computed wrong in Robolectric when running in Java 8/11, but not 9 (though I'm not sure yet why). Specifically, after submitting 2 wrong answers the following children should be in the state recycler view:
- ContentViewModel
- PreviousResponsesHeaderViewModel
- SubmittedAnswerViewModel
- FeedbackViewModel
- FractionInteractionViewModel
- SubmitButtonViewModel

This is 6 total children, but the view group only has 5 (except on Java 9 where it's correctly 6). For both Java 8 and 9 the adapter data list & item counts are correct, it's just the view group that's missing something.

BenHenning commented:
Ah, current theory is that something's different about viewport computation (leading to an unreliable # of views being cut-off in the viewport). That shouldn't be happening, but maybe there's a JDK 9-specific regression?

BenHenning commented:
So it turns out that for reasons not completely clear part of the content view model is still in the viewport when running on Java  9 (only 22 pixels). However, this is enough to make the child count check fail. While I don't understand *why* that's happening, it's a good indicator that doing a child count comparison is pretty fragile. Swapping it out with a more reliable check (e.g. counting the number of submitted answer view models) seems preferable.

BenHenning committed:
Fix timing issues and JDK 9-specific regression.

See comment thread in #1904 for more context.

BenHenning committed:
Restore workflow names + introduce test check.

The new test check workflow will be a static job that will be required
to pass. One failing test is being introduced to verify that this check
fails as expected.

The original workflow names are being restored so that they don't need
to be renamed in GitHub settings (since that would introduce a
discontinuity in CI service & require multiple migratiaon PRs to fix).

BenHenning commented:
This latest push will be used to verify:
- [x] 1. That the original workflow names match the required list
- [x] 2. That the new summarized status checker works as expected (fails)
- [x] 3. That the fixes to StateFragmentLocalTest work on CI

BenHenning committed:
Update StateFragmentLocalTest.kt

Remove fail-on-purpose test since it verified what it needed to: the new test status check job is working as expected.

BenHenning commented:
Okay, looks like everything is working as expected. Running one more CI without the fail-on-purpose check in place before sending this out for review.

BenHenning commented:
One flaky test is failing, but it's unrelated to this PR (first time I saw this fail today).

BenHenning commented:
Oops, didn't mean to close this. @anandwana001 & @vinitamurthi PTAL.

Also, you may want to wait until the latest test pass finishes (it's re-running since I accidentally closed the branch).

BenHenning commented:
Moving the discussion on build reproducibility to #1861.

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning committed:
Merge branch 'build-all-targets-in-bazel-ci' of github.com:oppia/oppia-android into build-all-targets-in-bazel-ci

BenHenning commented:
@anandwana001 and @vinitamurthi I am a bit eager to get this merged so that we can have guards against Bazel targets breaking. Could one of you maybe take a look before EOW?

Thanks.

anandwana001 commented on .bazelrc:
```diff
@@ -0,0 +1,9 @@
+# Configurations for arguments that should automatically be added to Bazel commands.
```
As per point number 8 in the PR description, are we planning to introduce `.bazelignore` and moving .gitignore out?

```
This PR also introduces updates to .gitignore and a new .bazelrc file to simplify working with 
Bazel. The latter will be iterated on as we approach moving the entire team over to Bazel 
permanently.
```

BenHenning replied:
Do we actually need anything in ``.bazelignore`` at the moment? My point was just meaning that we'll add more to both ``.bazelrc`` and ``.gitignore`` as we move over to Bazel permanently--given that everything builds, I don't think anything needs to be ignored by Bazel currently.

anandwana001 commented on .github/CODEOWNERS:
```diff
@@ -11,6 +11,9 @@
 #   the above date passes.
 
+# GitHub Actions & CI workflows.
+.github/*.yaml @BenHenning
```
Does `.yaml` and `.yml` works both with this?

vinitamurthi replied:
+1 same question, we should probably add .yml if it doesn't

BenHenning replied:
Good point! Added.

anandwana001 commented on .github/workflows/main.yml:
```diff
@@ -140,33 +140,172 @@ jobs:
+          GIT_SECRET_GPG_PRIVATE_KEY: ${{ secrets.GIT_SECRET_GPG_PRIVATE_KEY }}
+        run: |
+          cd $HOME
+          # NOTE TO DEVELOPERS: Make sure to never print this key directly to stdout!
```
It looks like this line needs a tab back to be a comment, right now I guess it's not working as a comment. Or it's just the IDE!!

<img width="907" alt="Screenshot 2020-12-04 at 14 35 10" src="https://user-images.githubusercontent.com/25057618/101144187-f0747100-363d-11eb-956c-2469468ae4be.png">
<img width="909" alt="Screenshot 2020-12-04 at 14 35 01" src="https://user-images.githubusercontent.com/25057618/101144198-f4a08e80-363d-11eb-810e-a426a79af05a.png">

BenHenning replied:
It actually is being "executed" as a comment as far as I can tell (the ``run`` block just executes everything in a shell, and ``#`` conventionally is ignored by shells). I actually lean toward keeping the code as-is because it's the echo line that's specifically sensitive and needs care taken when it's changed.

anandwana001 replied:
Ok

anandwana001 commented on .github/workflows/main.yml:
```diff
@@ -140,33 +140,172 @@ jobs:
+          cd $HOME/oppia-bazel
+          unzip bazel-build.zip
+          cd $GITHUB_WORKSPACE
+          chmod a+x $HOME/oppia-bazel/bazel
```
Do we have a modular way, as the cloning oppia/bazel and setting the secret key is something we are performing more than once?

BenHenning replied:
Yes, this was added to GitHub recently: https://github.blog/changelog/2020-08-07-github-actions-composite-run-steps/. It requires creating custom actions that perform sub-steps in the same repo. I didn't pursue it since it would take more time to figure out how to do this correctly, but I can look into if you feel strongly (I'm also not a fan of reusing large, complicated portions of the actions script, but I also feel that this is something we can simplify in subsequent PRs).

anandwana001 replied:
Sounds good. This could be a good learning issue around GitHub actions.

BenHenning replied:
Good idea--will keep this open as a reminder for me to file an issue for this later.

BenHenning replied:
Filed #2692.

anandwana001 commented on .github/workflows/workflow_canceller.yml:
```diff
@@ -0,0 +1,19 @@
+name: Automatic Workflow Canceller
+
+on:
```
We can add a top-level comment on when this action is going to get triggered.

BenHenning replied:
Added some details around this.

anandwana001 commented on .github/workflows/workflow_canceller.yml:
```diff
@@ -0,0 +1,19 @@
+    name: Cancel Previous Runs
+    runs-on: ubuntu-18.04
+    steps:
+      - uses: styfle/cancel-workflow-action@0.6.0
```
Should we mention the reference? Might be a learning curve
https://github.com/styfle/cancel-workflow-action

BenHenning replied:
Sure. I sort of expect people to be able to find these actions directly, but the additional comment doesn't hurt.

anandwana001 commented on app/src/main/res/layout-land/submitted_answer_item.xml:
```diff
@@ -20,6 +20,7 @@
   </data>
 
   <androidx.constraintlayout.widget.ConstraintLayout
+    android:id="@+id/submitted_answer_container"
```
Should we add the same in its tablet related UI file?

BenHenning replied:
This is apparently already defined for tablets: https://github.com/oppia/oppia-android/blob/develop/app/src/main/res/layout-sw600dp-land/submitted_answer_item.xml#L67. This PR is making the ID consistent across the layout files.

anandwana001 commented on scripts/compute_affected_tests.sh:
```diff
@@ -0,0 +1,77 @@
+
+  # Filter all of the source files among those that are actually included in Bazel builds.
+  changed_bazel_files=()
+  for changed_file in ${changed_files[@]}; do
```
Should we put in a double quote?

```console
for changed_file in "${changed_files[@]}"; do
```

anandwana001 replied:
Same with line number 47 and 54.

BenHenning replied:
I think this only matters if file names or directories contain spaces which won't be valid in the pure Bazel world (since the directory structure needs to match Java/Kotlin package structures & it's not valid for there to be spaces), but it makes the script a bit more robust so I went ahead and added them. Thanks!

anandwana001 commented on scripts/compute_affected_tests.sh:
```diff
@@ -0,0 +1,77 @@
+  changed_bazel_support_files=()
+  for changed_file in ${changed_files[@]}; do
+    if [[ "$changed_file" =~ ^.+?\.bazel$ ]] || [[ "$changed_file" =~ ^.+?\.bzl$ ]] || [[ "$changed_file" == "WORKSPACE" ]]; then
+      changed_bazel_support_files+=($changed_file)
```
```console
changed_bazel_support_files+=("$changed_file")
```

BenHenning replied:
Ditto & done.

anandwana001 reviewed (changes requested):
Tried to understand a bit of it, checked files and it looks like this is something we are really in need. 
Didn't get how the automatic workflow cancellation gets triggered!!

vinitamurthi commented on .github/workflows/main.yml:
```diff
@@ -140,33 +140,172 @@ jobs:
+          unzip bazel-tools.zip
+          tar -xf $HOME/oppia-bazel/android_tools.tar.gz -C $GITHUB_WORKSPACE/tmp/android_tools
+      # See explanation in bazel_build_app for how this is installed.
+      - name: Install git-secret (non-fork only)
```
This step is being used in multiple places, can we maybe define the step (or the block of code ) once and re-use it below?

BenHenning replied:
See my reply to @anandwana001's comment above--it's not straightforward to reuse stuff in actions (unless it can be put in a script, and we need to be really careful with the secret so that we don't accidentally print it in plaintext in the Bash history file or as part of the process command line visible via ``ps``.

vinitamurthi reviewed (commented):
General question, should we split these set of tests from the original CI and Lint tests? Or alternatively call the suite by a different name? What do you think?

BenHenning commented:
Thanks for the reviews! PTAL at the follow-up changes.

@anandwana001: maybe skim through https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows. That should help better understand the ``on`` construct in actions. In general, the cancellation workflow always gets run at the same time as lint tests one (and doesn't need to be done in an ordered way since it will only ever change the previous workflow runs).

@vinitamurthi: I think we should rename the workflows & split them up, but it will break the presubmit check (since GitHub hardcodes these). I suggest we do that after this PR gets submitted so that it can be done in isolation.

BenHenning committed:
Address reviewer comments.

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning commented:
@vinitamurthi @anandwana001 friendly ping to please take a look again. :)

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning committed:
Fix most tests broken in Bazel after syncing.

BenHenning committed:
Gitignore + fix broken test.

The test failure here only happens when using JDK9+ (which doesn't
happen in the Gradle world, or on CI).

The .gitignore is since we can't yet specify a .bazelproject in a
shareable way.

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning committed:
Lint fixes.

BenHenning commented:
I just noticed that GitHub can render the dependencies for actions which is really useful for us now that the workflow is starting to get much more complex: https://github.com/oppia/oppia-android/actions/runs/424948426.

fsharpasharp commented:
I'll take a look at this.

BenHenning commented:
Thanks @fsharpasharp! Added.

anandwana001 commented on domain/src/main/java/org/oppia/android/domain/classify/rules/numericinput/NumericInputIsGreaterThanRuleClassifierProvider.kt:
```diff
@@ -11,7 +11,8 @@ import javax.inject.Inject
  * https://github.com/oppia/oppia/blob/37285a/extensions/interactions/NumericInput/directives/numeric-input-rules.service.ts#L27
  */
-internal class NumericInputIsGreaterThanRuleClassifierProvider @Inject constructor(
+// TODO(#1580): Re-restrict access using Bazel visibilities
```
adding this a reference to check before merge if all the files have this todo because till the time it merge, new files are merging from other PR's.

BenHenning replied:
I checked this before I sent the PR back for review, but I'll double check before merging.

fsharpasharp commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -298,8 +305,15 @@ class StateFragmentLocalTest {
+      // The previous response header and only the last failed answer should be showing (since the
+      // failed answer list is collapsed).
       onView(withId(R.id.previous_response_header)).check(matches(isDisplayed()))
-      onView(withId(R.id.state_recycler_view)).check(matches(hasChildCount(/* childCount= */ 5)))
```
Could there any value in keeping this assertion. In case there's an unintended change to the collaped list? Or is it better to assert on what's actually visible in these tests?

fsharpasharp commented on domain/BUILD.bazel:
```diff
@@ -10,7 +10,10 @@ load("//domain:domain_test.bzl", "domain_test")
     name = "domain",
-    srcs = glob(["src/main/java/org/oppia/android/domain/**/*.kt"]),
+    srcs = glob(
+        ["src/main/java/org/oppia/android/domain/**/*.kt"],
```
Could we replace the glob in future with explicit dependencies? Or is it better to wildcard glob since we always want all of the non-test files in this target?

BenHenning replied:
Good question--it's definitely better to be specific. The reason we aren't here is because we explicitly need domain to include everything (in case new files are added). Phase 2 of the Bazel migration involves splitting this up into smaller, directory-based modules. We wouldn't use globs for those which means long-term we wouldn't use wildcards anymore.

fsharpasharp commented on domain/src/main/java/org/oppia/android/domain/classify/rules/fractioninput/FractionInputHasFractionalPartExactlyEqualToRuleClassifierProvider.kt:
```diff
@@ -13,7 +13,8 @@ import javax.inject.Inject
  *
  * https://github.com/oppia/oppia/blob/37285a/extensions/interactions/FractionInput/directives/fraction-input-rules.service.ts#L61
  */
-internal class FractionInputHasFractionalPartExactlyEqualToRuleClassifierProvider
```
I see that there'a TODO to restrict access, but is there a reason we can't keep internal modifier for now? Or are we expecting more uses of this outside its package?

BenHenning replied:
We actually can't use the internal modifier at all here. 'internal' means that only things within the same _build unit_ can access it, but the test accessing this provider is being built in a separate build unit. We can, however, use Bazel's own visibility system for this (e.g. by restricting access to the provider only to other packages that need it & tests). The TODO is tracking using Bazel visibilities, instead.

fsharpasharp reviewed (commented):
My comments are coming from me trying to understand the context, not from being familiar with the intended changes.

pubring.kbx~ should be removed I think. 

I also agree about the re-using the secret steps, but I'm aware that you already addressed that.

BenHenning commented:
@anandwana001 what other action items do I need to address per your requested changes?

anandwana001 reviewed (approved):
LGTM

BenHenning commented:
Thanks @anandwana001, @vinitamurthi, and @fsharpasharp!

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

Conflicts:
	domain/src/test/java/org/oppia/android/domain/oppialogger/OppiaLoggerTest.kt

BenHenning committed:
Post-merge clean-up.

BenHenning commented:
I think the StateFragmentTest failures *might* be real images taking too long to load (or failing to load) in the test environment. We really shouldn't be loading real images at all, and may need to address that if it continues to cause Bazel-CI flakes.

BenHenning commented:
Closing & reopening to re-trigger GitHub actions since they weren't automatically triggering.

BenHenning committed:
Fix broken post-merge test.

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning committed:
Remove unnecessary codeowners per earlier codeowners setup.

BenHenning committed:
Fix
ItemSelectionInputDoesNotContainAtLeastOneOfRuleClassifierProviderTest.

BenHenning committed:
Disable image region selection tests.

These tests can't correctly pass on Robolectric until #1611 is fixed, so
disabling them for the time being to avoid the image loading flake
happening on CI (but not locally). Note that chances are a fix will
still be needed for the flake, but that can be addressed later.

BenHenning commented:
Fixed a few more aspects of this PR, including:
- Fixing ItemSelectionInputDoesNotContainAtLeastOneOfRuleClassifierProviderTest
- Disabling image region selection tests. I'm not sure why they're failing only on CI with Bazel (and weren't earlier), but I think it's reasonable to disable them until #1611 can be directly addressed. If they pass now, that's technically an incorrect pass since the render node computations are incorrect with Robolectric.
- Removing unnecessary codeowners changes since the current develop codeowners file is more correct and encapsulates the original codeowners changes being introduced by this PR

BenHenning committed:
Disable 2 previously missed tests.

BenHenning commented:
Weird. Lots of CI network issues it seems.

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

Conflicts:
	app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt

BenHenning committed:
Post-merge lint fix.

BenHenning committed:
Add missing dependency.

Verified all tests build & pass both for JDK 9 & 11. Hopefully they will
work as expected on CI.

BenHenning committed:
Add missing codeowners for new files.

BenHenning commented:
Welp 1 test seems to be hanging. This is sort of the worst-case scenario being demonstrated (one failure/hang/flake triggering a need to re-run the whole suite). I think I might need to introduce a custom check to re-run only failing tests, but it's not obvious how to do this in a way that makes GitHub happy. I'll need to figure out if I can easily trigger a re-run from a GitHub comment, and somehow only run the failing tests (which may require some special cache upload/download).

BenHenning commented:
Re-triggering. I think we can merge this in as-is, but we may not be able to enable the check until we have a better idea how likely transient failures are.

BenHenning commented:
@rt4914 PTAL for codeowners.

rt4914 reviewed (approved):
Approving for code-owner files.

BenHenning commented:
Need to re-run everything due to Gradle flake. :(

BenHenning commented:
@seanlip please approve codeowners change.

seanlip commented:
Done.

@BenHenning just to check, is `pubring.kbx~` (with the tilde at the end) a file you intended to check in? It looks like some sort of backup so I thought I'd double-check.

BenHenning commented:
@seanlip so much as I can tell that file is expected to be present. I don't totally understand how git secret manages it internally, but it doesn't gitignore that file (which I would expect if it weren't necessary since it does gitignore others). It may be a necessary backup for the inner workings of git secret, not sure.

Thanks for double checking!

BenHenning committed:
Merge branch 'develop' into build-all-targets-in-bazel-ci

Conflicts:
	.github/workflows/main.yml
	domain/BUILD.bazel
	testing/BUILD.bazel

BenHenning commented:
Note that I'm going to also split the Bazel tests into their own workflow so that failures don't require restarting all checks (only the Bazel ones). I've also filed the post-PR tracking issues now in preparation of merging this shortly.

BenHenning commented:
I think we also need a way to guarantee all tests are covered as we continue with the migration, otherwise it seems a bit too easy to miss things. Will follow up with something in a PR after this one.

BenHenning committed:
Post-merge fixes.

This fixes some tests that were broken after recent PRs, and fixed a
visibility error introduced in #2663.

BenHenning committed:
Move Bazel tests to new workflow.

This will make it easier to restart failures without having to also
restart unrelated tests.

BenHenning commented:
I think I'm also going to move this to using Bazel 4.0 LTS since it will substantially speed up the cloning part of the workflow.

BenHenning commented:
Awesome, green PR. :) Going to merge this in as-is & add the optimization in a follow-up PR since this unlocks being able to fix a few different things.

BenHenning commented:
Thanks all for the feedback & reviews. This was a bit more complicated of a PR, and took quite a bit of time to get it merge-worthy.
